sdk/Trace/Span.php

```diff
@@ -59,383 +81,374 @@ public static function fromContext(Context $context): API\Span
+                $line ?? ''
+            );
+            $seen[] = "$file:$line";
+            if (!count($trace)) {
+                break;
+            }
+            $file = array_key_exists('file', $trace[0]) ? $trace[0]['file'] : 'Unknown Source';
+            $line = array_key_exists('file', $trace[0]) && array_key_exists('line', $trace[0]) && $trace[0]['line'] ? $trace[0]['line'] : null;
+            array_shift($trace);
+        }
+        $result = implode("\n", $result);
+        if ($prev) {
+            $result  .= "\n" . self::formatStackTrace($prev, $seen);
+        }
 
-        // todo: set these to null until needed
-        $this->attributes = $attributes ?? new Attributes();
-        $this->events = new Events();
-        $this->links = $links ?? new Links();
+        return $result;
     }
 
-    /**
-     * @internal
-     */
-    public function setInstrumentationLibrary(InstrumentationLibrary $instrumentationLibrary)
+    /** @inheritDoc */
+    public static function wrap(API\SpanContext $spanContext): API\Span
     {
-        $this->instrumentationLibrary = $instrumentationLibrary;
+        return new NonRecordingSpan($spanContext);
     }
 
-    /**
-     * @internal
-     */
-    public function getInstrumentationLibrary(): InstrumentationLibrary
-    {
-        return $this->instrumentationLibrary;
-    }
+    // TODO: Add a SpanLimits object to the constructor for configuration options.
 
-    public function getResource(): ResourceInfo
-    {
-        return clone $this->resource;
-    }
+    /** @readonly */
+    private API\SpanContext $context;
 
-    public function getParent(): ?API\SpanContext
-    {
-        // todo: Spec says a parent is a Span, SpanContext, or null -> should we implement this here?
-        return $this->parentSpanContext !== null ? clone $this->parentSpanContext : null;
-    }
+    /** @readonly */
+    private API\SpanContext $parentSpanContext;
+
+    /** @readonly */
+    private SpanProcessor $spanProcessor;
 
     /**
-     * @param string $code
-     * @param string|null $description
-     * @return Span
+     * @readonly
+     *
+     * @todo: Java just has this as list<API\Event>, could we just do that?
      */
-    public function setSpanStatus(string $code, ?string $description = null): API\Span
-    {
-        if ($this->isRecording()) {
-            $this->spanStatus->setStatus($code, $description);
-        }
+    private API\Events $events;
 
-        return $this;
-    }
+    /**
+     * @readonly
+     *
+     * @todo: Java just has this as list<API\Link>, could we just do that?
+     */
+    private API\Links $links;
 
-    public function getStart(): int
-    {
-        return $this->start;
-    }
+    /** @readonly */
+    private int $totalRecordedLinks;
 
-    public function setStart(int $start): Span
-    {
-        $this->start = $start;
+    /** @readonly */
+    private int $kind;
 
-        return $this;
-    }
+    // TODO: Store a clock instance on the Span?
```

Comment:
Java seems to store a `Clock` instance within the object, not sure there's any reason to do that in our case given you can just `Clock::get()`. In Java's case I think it's special in that the clock is anchored to when the span is created, versus just reading system time info every time.

Reply:
Yeah I think that our implementation is good!